README.md
r-conomic
=========

Ruby wrapper for the [e-conomic](http://www.e-conomic.co.uk) SOAP API, that aims at making working with the API bearable.

E-conomic is a web-based accounting system. For their marketing speak, see [http://www.e-conomic.co.uk/about/](http://www.e-conomic.co.uk/about/). More details about their API at [http://www.e-conomic.com/developer](http://www.e-conomic.com/developer).

[![Build Status](https://travis-ci.org/substancelab/rconomic.svg?branch=master)](https://travis-ci.org/substancelab/rconomic) [![Test Coverage](https://codeclimate.com/github/substancelab/rconomic/badges/coverage.svg)](https://codeclimate.com/github/substancelab/rconomic/coverage) [![Code Climate](https://codeclimate.com/github/substancelab/rconomic/badges/gpa.svg)](https://codeclimate.com/github/substancelab/rconomic)


Usage example
-------------

    economic = Economic::Session.new

    # Connect using a Private app ID and an access ID provided by the "Grant Access"
    # As described here: https://www.e-conomic.com/developer/connect
    economic = Economic::Session.new
    economic.connect_with_token 'the_private_app_id', 'the_access_id_you_got_from_the_grant'

    # Find a debtor:
    debtor = economic.debtors.find(:number => 101)

    # Creating a debtor:
    debtor = economic.debtors.build

    debtor.number = economic.debtors.next_available_number
    debtor.debtor_group_handle = { :number => 1 }
    debtor.name = 'Bob'
    debtor.vat_zone = 'HomeCountry' # HomeCountry, EU, Abroad
    debtor.currency_handle = { :code => 'DKK' }
    debtor.price_group_handle = { :number => 1 }
    debtor.is_accessible = true
    debtor.ci_number = '12345678'
    debtor.term_of_payment_handle = { :id => 1 }
    debtor.layout_handle = { :id => 16 }
    debtor.save

    # Create invoice for debtor:
    current_invoice = debtor.current_invoices.build
    current_invoice.term_of_payment_handle = debtor.term_of_payment_handle

    current_invoice.date = Time.now
    current_invoice.due_date = Time.now + 15
    current_invoice.exchange_rate = 100
    current_invoice.is_vat_included = false

    invoice_line = Economic::CurrentInvoiceLine.new
    invoice_line.description = 'Line on invoice'
    invoice_line.unit_handle = { :number => 1 }
    invoice_line.product_handle = { :number => 101 }
    invoice_line.quantity = 12
    invoice_line.unit_net_price = 19.95
    current_invoice.lines << invoice_line

    current_invoice.save

    # You can delete it by doing:
    # current_invoice.destroy

    invoice = current_invoice.book

    # Create a debtor payment

    cash_book = economic.cash_books.all.last # Or find it by its number

    # The reason debtor payments are done this way is because we don't want to specify the voucher number. If we build the cash book entry ourselves,
    # without specifying the voucher number, the API will complain. This way, E-Conomics will assign a voucher number for us.

    cash_book_entry = cash_book.entries.create_debtor_payment(:debtor_handle => debtor.handle, :contra_account_handle => { :number => '1920' })
    cash_book_entry.amount = -123.45
    cash_book_entry.currency_handle = { "Code" => "DKK" }
    cash_book_entry.debtor_invoice_number = invoice.number
    cash_book_entry.text = "Payment, invoice #{ invoice.number }"
    cash_book_entry.save

    cash_book.book

    # Find a product:
    product = economic.products.find(1234)

    # Creating a product:
    product = economic.products.build
    product.number = 'ESC2014-LED-DISPLAY'
    product.product_group_handle = { :number => 1 }
    product.name = '100 meter LED display'
    product.sales_price = 999999
    product.cost_price = 100000
    product.recommended_price = 999999
    product.is_accessible = true
    product.volume = 1
    product.save

How to enable e-conomic API access
----------------------------------

You need a Developer account and setup an app in their web UI. E-conomic users can then grant that app access to their account.

* Developer portal: https://www.e-conomic.com/developer
* Instructions: https://www.e-conomic.com/developer/connect

It doesn't do everything
------------------------

Not even remotely... For now, limited to a small subset of all the [available operations](https://www.e-conomic.com/secure/api1/EconomicWebService.asmx):

                       | Create | Read | Update | Delete
    -------------------+--------+------+--------+-------
    CashBook           | X      | X    | X      | X
    CashBookEntry      | X      | X    | X      | X
    Company            | X      | X    | X      | X
    Creditor           | X      | X    | X      | X
    CreditorEntry      | X      | X    |        |
    CreditorContact    | X      | X    | X      | X
    CreditorEntry      | X      | X    | X      | X
    CurrentInvoice     | X      | X    | X      | X
    CurrentInvoiceLine | X      | X    | X      | X
    Debtor             | X      | X    | X      | X
    DebtorContact      | X      | X    | X      | X
    DebtorEntry        | X      | X    | X      | X
    Entry              | X      | X    | X      | X
    Invoice            | X      | X    |        |
    Order              |        | X    |        |
    OrderLine          |        | X    | X      |
    Product            | X      | X    | X      | X

Credits
-------

Sponsored by [Lokalebasen.dk](http://lokalebasen.dk)


License
-------

R-conomic is licensed under the MIT license. See LICENSE for details.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Added

* `CurrentInvoice` can now fetch and return it's PDF representation, similar to `Invoice` (@olepalm).
* Added `EntityProxy#get_data_for_handles` that makes it possible to fetch data for a whole list of entities at once (@olepalm).
* It is now possible to add `DebtorContact`s to `CurrentInvoice#your_reference`.

### Changes

*

### Removed

*

## 0.7.2 [June 24, 2019]

### Added

* Ability to create ManualDebtorInvoice cash book entries (@prognostikos).

### Changes

* `Debtor#current_invoices` now only returns invoices for the debtor (@olepalm).

## 0.7.1 [February 21st, 2018]

### Added

* Operation to fetch only current orders from the backend.
* Support for OrderLine objects (@olepalm)
* Ability to get order lines for a specific order (@olepalm).
* Orders can now be found by their other_reference field (@olepalm).
* Ability to toggle Sent-status of orders (@olepalm).

### Changes

* The cached E-conomic WSDL has been updated with the most recent version.

### Removed

## 0.7.0 [July 1st, 2018]

This release contains breaking changes as E-conomic has deprecated one of their connection methods. See https://www.e-conomic.com/developer#_ga=2.79896535.1117342868.1530474675-2146547048.1526554784 for details.

### Removed

* Ability to connect with agreement number, username and password. You need to use `Economic::Session#connect_with_token` going forward.
* Ability to set an app identifier. This was tied to the now removed `Economic::Session#connect_with_credentials` method.

## 0.6.1 [July 1st, 2018]

### Changes

* Tighten Savon dependency. Due to changes in Savons internal API we cannot work with versions 2.11.2 or later.

## 0.6.0 [April 30, 2017]

### Added

* Support for getting all contacts for a given debtor (@nielsbuus)
* Basic implementation of products (@traels)
* Ability to create CreditorInvoice entries (@prognostikos)
* Properties for "past due" status of invoices (@skelboe)
* Support to get company data (@adriacidre)
* Support X-EconomicAppIdentifier (@prognostikos)
* Fixed an issue where X-EconomicAppIdentifier wasn't set properly (@olepalm)

### Changes

* The source code has a new home at https://github.com/substancelab/rconomic
* Non-existing methods removed from the documentation (@koppen)
* Handle Numbers can now be String as well as numbers (@traels)
* Fixed authentication when using legacy Connect (@koppen)

### Removed

* We no longer officially support Ruby 1.9.3. rconomic might still work on
  legacy rubies, but we aren't testing it.


lib/economic/product.rb
# frozen_string_literal: true

require "economic/entity"

module Economic
  # Represents a product in E-conomic.
  #
  # API documentation: http://www.e-conomic.com/apidocs/Documentation/T_Economic_Api_IProduct.html
  #
  # Examples
  #
  #   # Find a product:
  #   product = economic.products.find(1234)
  #
  #   # Creating a product:
  #   product = economic.products.build
  #   product.number = 'ESC2014-LED-DISPLAY'
  #   product.product_group_handle = { :number => 1 }
  #   product.name = '100 meter LED display'
  #   product.sales_price = 999999
  #   product.cost_price = 100000
  #   product.recommended_price = 999999
  #   product.is_accessible = true
  #   product.volume = 1
  #   product.save
  class Product < Entity
    has_properties :number,
      :product_group_handle,
      :name,
      :description,
      :bar_code,
      :sales_price,
      :cost_price,
      :recommended_price,
      :unit_handle,
      :is_accessible,
      :volume,
      :department_handle,
      :distribution_key_handle,
      :in_stock,
      :on_order,
      :ordered,
      :available

    def handle
      @handle ||= Handle.new(number: @number)
    end

    protected

    def fields
      to_hash = proc { |handle| handle.to_hash }
      [
        ["Handle", :handle, to_hash, :required],
        ["Number", :handle, proc { |h| h.number }, :required],
        ["ProductGroupHandle", :product_group_handle, to_hash],
        ["Name", :name, nil, :required],
        ["Description", :description, nil],
        ["BarCode", :bar_code, nil],
        ["SalesPrice", :sales_price, nil, :required],
        ["CostPrice", :cost_price, nil, :required],
        ["RecommendedPrice", :recommended_price, nil, :required],
        ["UnitHandle", :unit_handle, to_hash],
        ["IsAccessible", :is_accessible, nil, :required],
        ["Volume", :volume, nil, :required],
        ["DepartmentHandle", :department_handle, to_hash],
        ["DistributionKeyHandle", :distribution_key_handle, to_hash],
        ["InStock", :in_stock],
        ["OnOrder", :on_order],
        ["Ordered", :ordered],
        ["Available", :available]
      ]
    end
  end
end
